### ISSUE TITLE
Class decorator

### ISSUE THREAD

**Issue by janvainer** [pseeth/argbind#4]
Hi, awesome work! Its really nice and simple :) Did you consider allowing class init? It could be really nice for initialization of eg. NN models:

```
@argbind.bind()
class NeuralNet(nn.Module):

    def __init__(
        self,
        n_layers: int = 5,
        use_batch_norm: bool = True,
        in_channels: int = 80,
    ):
    """Docstring with help"""
        ...
...
with argbind.scope(args):
    nn = NeuralNet()

# or
nn = NeuralNet(**args["NeuralNet"])  # this would work if the dict structure would not be flat, but nested, which should not be a big issue
```
Right now, to instantiate a class it is necessary to write a constructor function. Something like:
```
def build_neural_net(#args go here):
    """Docstring with help"
    nn = NeuralNet(#pass args)
    return nn
```

What do you think?

**Comment by pseeth** [comment]
Thanks! I think putting better class decoration can definitely be on the roadmap. For now, though, what I do is add a function to my classes that's like this:

```python
import argbind

class NeuralNet(nn.Module):
  def __init__(
      self,
      n_layers: int = 5,
      use_batch_norm: bool = True,
      in_channels: int = 80,
  ):
  """Docstring with help"""
      pass

  @classmethod
  def bind(cls):
      return argbind.bind()(cls)
```

Then, you can get the "bound" class function via:

```python
neural_net = NeuralNet.bind()
args = argbind.parse_args()
with argbind.scope(args):
  neural_net()
```

and it's parameters are available like so:

```bash
usage: neural_net.py [-h] [--args.save ARGS.SAVE] [--args.load ARGS.LOAD] [--args.debug ARGS.DEBUG] [--NeuralNet.n_layers NEURALNET.N_LAYERS] [--NeuralNet.use_batch_norm] [--NeuralNet.in_channels NEURALNET.IN_CHANNELS]

optional arguments:
  -h, --help            show this help message and exit
  --args.save ARGS.SAVE
                        Path to save all arguments used to run script to.
  --args.load ARGS.LOAD
                        Path to load arguments from, stored as a .yml file.
  --args.debug ARGS.DEBUG
                        Print arguments as they are passed to each function.

Generated arguments for function NeuralNet:

  --NeuralNet.n_layers NEURALNET.N_LAYERS
  --NeuralNet.use_batch_norm
  --NeuralNet.in_channels NEURALNET.IN_CHANNELS
```

Supporting this more officially I can definitely look into when I get some time!

As far as supporting nested args goes, that likely won't be on the roadmap. I think it might add a bit too much complexity. The idea is to keep the `.yml` file syntax as close as possible to the `argparse` syntax, so that the mental model of how to launch and configure your program is exactly the same. Since nested args aren't very clean to do from CLI (at least I can't think of a way to do it cleanly), then it probably won't happen.

**Comment by janvainer** [comment]
Thanks for the reply. The classmethod binding looks cool.  :) Regarding the nesting, the thing is that the args namespace is basically a flattened dict that originally had one level of nesting:
```
function.arg1 = 0
function.arg2 = 1
...
# the one above is basically like the following yaml (maybe a bit easier to read if config is long). I would not use more than a single level of nesting
function: 
    arg1: 0
    arg2: 1
```

I think my main point is regarding nesting is to allow something like dot-notation access to the attributes of a single function and scoped unpacking:
```
>>> args.NeuralNet  # returns a dict or some namespace containing only NeuralNet inputs
```
This would among other things allow easy use of args collected by bind methods without using the argbind scope.
```
nn = NeuralNet(**args.NeuralNet)
```
I am just brainstorming some ways how `argbind` could be used :)

**Comment by pseeth** [comment]
That's quite interesting. I think one easy way to do this would be to transparently change the underlying dictionary that contains all of the args so that it is composed of Namespaces for each function that was bound. But to do it in such a way that both dict-style and dot-style accesses are valid, so that it all stays backwards compatible. I'll have to think a bit more about this!

Thanks for using it! I hope it's helpful, anyway.

**Comment by janvainer** [comment]
Cool :) Let me know how things are going. I am trying to integrate this in one of my projects. Its really nice so far :)

**Comment by pseeth** [comment]
@janvainer , I addressed the first bit of this issue in #5. You can just call `argbind.bind` directly on a class to get a bound version of that class. This actually worked before too, but now it's easier to use. I'll add conversions to NameSpaces or different lookup abilities in a different PR down the road. Here's an example of the usage:

```python
import argbind

class NeuralNet:
  def __init__(
      self,
      n_layers: int = 5,
      use_batch_norm: bool = True,
      in_channels: int = 80,
  ):
    """Docstring with help"""
    pass

NeuralNet = argbind.bind(NeuralNet)
args = argbind.parse_args()
args['args.debug'] = True
with argbind.scope(args):
  NeuralNet()
```

```bash
❯ python scratch.py -h
usage: scratch.py [-h] [--args.save ARGS.SAVE] [--args.load ARGS.LOAD] [--args.debug ARGS.DEBUG] [--NeuralNet.n_layers NEURALNET.N_LAYERS] [--NeuralNet.use_batch_norm]
                  [--NeuralNet.in_channels NEURALNET.IN_CHANNELS]

optional arguments:
  -h, --help            show this help message and exit
  --args.save ARGS.SAVE
                        Path to save all arguments used to run script to.
  --args.load ARGS.LOAD
                        Path to load arguments from, stored as a .yml file.
  --args.debug ARGS.DEBUG
                        Print arguments as they are passed to each function.

Generated arguments for function NeuralNet:

  --NeuralNet.n_layers NEURALNET.N_LAYERS
  --NeuralNet.use_batch_norm
  --NeuralNet.in_channels NEURALNET.IN_CHANNELS
```

and usage:

```bash
❯ python scratch.py --NeuralNet.n_layers=200
NeuralNet(
  n_layers : int = 200
  use_batch_norm : bool = False
  in_channels : int = 80
)
```

It's useful for only binding what you need within a training script, and leaving your main package code ArgBind-less, which I think addresses a limitation of ArgBind now!

**Comment by janvainer** [comment]
Cool! its looking really good. :)

**Comment by pseeth** [comment]
I believe this is now fully addressed via #7, so I am closing! ArgBind now modifies the class's init function, and then returns the class. This keeps the class intact completely.
